File: Union-Find/uf.cpp
#include <vector>
using namespace std;


class Union_Find {
public:
    Union_Find(int n) {
        for (int i = 0; i < n; ++i) {
            par.push_back(i);
            rank.push_back(0);
        }
    }
    ~Union_Find() {

    }

    void _union(int a, int b) {
        int pa = _find(a);
        int pb = _find(b);
        if (pa != pb) {
            if (rank[pa] < rank[pb]) {
                par[pb] = pa;
            } else if (rank[pa] == rank[pb]) {
                par[pb] = pa;
                rank[pa]++;
            } else {
                par[pa] = pb;
            }
        }
    }

    int _find(int a) {
        if (par[a] != a) {
            par[a] = _find(par[a]);
        }
        return par[a];
    }
private:
    vector<int> par, rank;
};

class Union_Find_lazy {
public:
    Union_Find_lazy(int n) {
        for (int i = 0; i < n; ++i) {
            par.push_back(i);
            rank.push_back(0);
        }
    }
    ~Union_Find_lazy() {

    }

    void _union(int a, int b) {
        int pa = _find(a);
        int pb = _find(b);
        par[pa] = pb;
    }

    int _find(int a) {
        if (par[a] != a) {
            par[a] = _find(par[a]);
        }
        return par[a];
    }
private:
    vector<int> par;
};

int main() {
    return 0;
};

File: Union-Find/uf.py

class Union_Find:
    def __init__(self, n) -> None:
        self.rank = [0] * n
        self.par = [i for i in range(n)]
    
    def union(self, a, b):
        pa = self.find(a)
        pb = self.find(b)
        if pa != pb:
            if self.rank[pa] < self.rank[pb]:
                self.par[pb] = pa
            elif self.rank[pa] == self.rank[pb]:
                self.par[pb] = pa
                self.rank[pa] += 1
            else:
                self.par[pa] = pb
        
    def find(self, a):
        if a != self.par[a]:
            self.par[a] = self.find(self.par[a])
        return self.par[a]

class Union_Find_lazy:
    def __init__(self, n) -> None:
        self.par = [i for i in range(n)]
    
    def union(self, a, b):
        pa = self.find(a)
        pb = self.find(b)
        if pa != pb:
            self.par[pa] = pb
        
    def find(self, a):
        if a != self.par[a]:
            self.par[a] = self.find(self.par[a])
        return self.par[a]